### ISSUE TITLE
test_AnnexRepo_commit failures on Travis

### ISSUE THREAD

**Issue by kyleam** [datalad/datalad#4773]
`test_AnnexRepo_commit` failed in two recent PRs (gh-4762 and gh-4768).  I think it's intermittent.

- https://travis-ci.org/github/datalad/datalad/jobs/712219198#L1996
- https://travis-ci.org/github/datalad/datalad/jobs/712340298#L1991

```
======================================================================
ERROR: datalad.support.tests.test_annexrepo.test_AnnexRepo_commit
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/datalad/tests/utils.py", line 1185, in newfunc
    raise exc_info[1]
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/datalad/tests/utils.py", line 1157, in newfunc
    ret = func(*args, **kwargs)
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/datalad/tests/utils.py", line 728, in newfunc
    return t(*(arg + (filename,)), **kw)
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/datalad/support/tests/test_annexrepo.py", line 786, in test_AnnexRepo_commit
    assert_raises(FileNotInRepositoryError, ds.commit, files="not-existing")
  File "/opt/python/3.7.1/lib/python3.7/unittest/case.py", line 743, in assertRaises
    return context.handle('assertRaises', args, kwargs)
  File "/opt/python/3.7.1/lib/python3.7/unittest/case.py", line 178, in handle
    callable_obj(*args, **kwargs)
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/datalad/support/gitrepo.py", line 1654, in commit
    stdin=None,
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/datalad/cmd.py", line 504, in run
    **results,
datalad.support.exceptions.CommandError: CommandError: 'git commit -m '[DATALAD] Recorded changes' -- not-existing' failed with exitcode 1 under /tmp/datalad_temp_test_AnnexRepo_commitmkv02v44
```

**Comment by kyleam** [comment]
I've spent a good amount of time trying to trigger this on my local machine and in a Xenial VM.  I don't recall seeing it until after we switched to the `_DL_ANNEX_INSTALL_SCENARIO="conda-forge 7.20190819"` setup, so my focus has been on that.  No luck.

The failing assertion of the test expects a `FileNotInRepositoryError` to be raised from here:

https://github.com/datalad/datalad/blob/55200cd3b23041978f6f7fa0a15924945e861e45/datalad/support/gitrepo.py#L1656-L1667

But a `CommandError` comes instead.  My best guess is that something about the `stderr`  capture by the runner is racy in this case.

I think this is failing regularly enough that we should catch the `CommandError` in the test and skip it as a known failure while keeping this issue open.

**Comment by yarikoptic** [comment]
so may be rerun (on travis) with `DATALAD_LOG_CMD_OUTPUTS=1` to actually see what/where(stdout/stderr) is returned for the failing run, which could explain why our matching of the outputs (which you love I know) fails and we just re-`raise` that CommandError

**Comment by kyleam** [comment]
> so may be rerun (on travis) with `DATALAD_LOG_CMD_OUTPUTS=1` to actually see what/where(stdout/stderr) is returned for the failing run

Thanks for the suggestion.  Certainly worth seeing if that reveals anything, and, with the rate we're seeing this failure, it shouldn't take too many rebuilds to trigger it in a job.

**Comment by kyleam** [comment]
> > so may be rerun (on travis) with `DATALAD_LOG_CMD_OUTPUTS=1` to actually see what/where(stdout/stderr) is returned for the failing run
> 
> Thanks for the suggestion. Certainly worth seeing if that reveals anything, and, with the rate we're seeing this failure, it shouldn't take too many rebuilds to trigger it in a job.

On top of the fix in gh-4796, I tried to trigger the failure under `DATALAD_LOG_CMD_OUTPUTS=1` (origin/scratch/annex-commit-failure-debug branch).  Initially I tested with the entire `datalad.support.tests.test_annexrepo` module, thinking that perhaps the failure was influenced by other tests, but with `DATALAD_LOG_CMD_OUTPUTS=1` and `DATALAD_LOG_LEVEL=5` (necessary to see the log messages) that exceeds the maximum output limits.  I then set `DATALAD_LOG_LEVEL=10` and moved some of the relevant log statements to 10, but that still hits the limit.  Restricting the run farther to `test_AnnexRepo_commit`, I haven't been able to trigger the failure yet.  Will try a few more times.

**Comment by yarikoptic** [comment]
> but with DATALAD_LOG_CMD_OUTPUTS=1 and DATALAD_LOG_LEVEL=5 (necessary to see the log messages) that exceeds the maximum output limits.

and that is without `-s` for nose?  Without `-s` it should only print captured output for the failing test.

Another trick (although I leave it to you completely either to try it at all -- I am ok to skip this damn test ;)) could be - to redirect output to a log file (specify filename in `DATALAD_LOG_TARGET=`) and then just grep/tail/whatever for the block of lines of  interest if testing fails (and then fail the run)

**Comment by kyleam** [comment]
> and that is without `-s` for nose? Without `-s` it should only print captured output for the failing test.

`-s` is only relevant for stdout, but either way, yes, that is without `-s` (our default .travis.yml invocation doesn't include it).

My understanding is that by default nose should capture logs as well (disabled via `--nologcapture`), but, at least with our test setup, I don't see logs captured when running the tests locally with `DATALAD_LOG_LEVEL` set, and travis builds show the same thing.  Without `DATALAD_LOG_LEVEL` set in the environment, logging output doesn't show up because `setup_package` sets the log level to 100.

**Comment by kyleam** [comment]
> My understanding is that by default nose should capture logs as well (disabled via `--nologcapture`), but, at least with our test setup, I don't see logs captured

Or rather it seems the nose does capture logs but that doesn't prevent datalad's logs from also hitting stderr when `DATALAD_LOG_LEVEL` is set.  Because the logs are captured, `DATALAD_LOG_TARGET=/dev/null` can avoid the flooding, while failures will still show the captured log messages.

I wasn't able to trigger the `test_AnnexRepo_commit` failure by running just the `test_annexrepo` module, but I triggered it when running the entire test suite.  Not sure if that's just coincidence.

Anyway, here's one run.

https://travis-ci.org/github/datalad/datalad/builds/715249596#L2242

This run is on top of a recent commit in maint (7ff697da2).  The only change touching outside of CI config files is the `CommandError.stderr` debug message that can be seen in the output below.

<details>
<summary>diff</summary>

```diff
diff --git a/datalad/support/gitrepo.py b/datalad/support/gitrepo.py
index 02f7203c6..d3bcf7fe1 100644
--- a/datalad/support/gitrepo.py
+++ b/datalad/support/gitrepo.py
@@ -1654,6 +1654,7 @@ def commit(self, msg=None, options=None, _datalad_msg=False, careless=True,
                     stdin=None,
                 )
         except CommandError as e:
+            lgr.debug("CommandError.stderr: %s", e.stderr)
             # real errors first
             if "did not match any file(s) known to git" in e.stderr:
                 raise FileNotInRepositoryError(
```

</details>

```
[...]
     1	datalad.cmd: DEBUG: Async run ['git', 'commit', '-m', '[DATALAD] Recorded changes', '--', 'untracked']
     2	datalad.cmd: DEBUG: Launching process ['git', 'commit', '-m', '[DATALAD] Recorded changes', '--', 'untracked']
     3	datalad.cmd: DEBUG: Process 30498 started
     4	datalad.cmd: DEBUG: Waiting for process 30498 to complete
     5	datalad.cmd: Level 5: Read 67 bytes from 30498[stderr]:
     6	datalad.cmd: Level 5: stderr| error: pathspec 'untracked' did not match any file(s) known to git
     7	|  
     8	datalad.cmd: DEBUG: Process 30498 exited with return code 1
     9	datalad.cmd: Level 8: Finished running ['git', 'commit', '-m', '[DATALAD] Recorded changes', '--', 'untracked'] with status 1
    10	datalad.gitrepo: DEBUG: CommandError.stderr: error: pathspec 'untracked' did not match any file(s) known to git
       
    11	datalad.gitrepo: DEBUG: Committing via direct call of git: ['git', 'commit', '-m', '[DATALAD] Recorded changes']
    12	datalad.cmd: DEBUG: Async run ['git', 'commit', '-m', '[DATALAD] Recorded changes', '--', 'not-existing']
    13	datalad.cmd: DEBUG: Launching process ['git', 'commit', '-m', '[DATALAD] Recorded changes', '--', 'not-existing']
    14	datalad.cmd: DEBUG: Process 30499 started
    15	datalad.cmd: DEBUG: Waiting for process 30499 to complete
    16	datalad.cmd: DEBUG: Process 30499 exited with return code 1
    17	datalad.cmd: Level 5: Read 70 bytes from 30499[stderr]:
    18	datalad.cmd: Level 5: stderr| error: pathspec 'not-existing' did not match any file(s) known to git
    19	|  
    20	datalad.cmd: Level 8: Finished running ['git', 'commit', '-m', '[DATALAD] Recorded changes', '--', 'not-existing'] with status 1
    21	datalad.gitrepo: DEBUG: CommandError.stderr: 
    22	datalad.utils: Level 5: Removing temp file: /tmp/datalad_temp_test_AnnexRepo_commitgqclb78q
```

`assert_raises(FileNotInRepositoryError, ds.commit, files="untracked")` fails as expected. `StdErrCapture` reports capturing its stderr (lineno 6) and that stderr makes it through to the `CommandError.stderr` that `commit` inspects (lineno 10).

The next assertion is this issue is about: `assert_raises(FileNotInRepositoryError, ds.commit, files="not-existing")`.  `StdErrCapture` again logs the expected message (lineno 18).  But it's lost somewhere and doesn't make it into `CommandError.stderr` (lineno 21).

So that's something to go on.

**Comment by yarikoptic** [comment]
uff, seems to be a Runner issue loosing some "tail" output!  

Unlikely of help so FWIW: I remember running into some cases where I needed "to read more" while I was trying to figure out those "new git annex stalls"... I think it might have been this PR https://github.com/datalad/datalad/pull/4277: where I have tried a few more things like https://github.com/datalad/datalad/pull/4277/commits/ea2c2cc318a5999a29939493bb70c9e4f85e37f6 ... 
but that PR doesn't look like the one where I remember dumping more notes... I might have mixed up what I troubleshooted when since https://github.com/datalad/datalad/pull/4265 looks like the one, but it lacks traces any tuning of the Runner

**Comment by kyleam** [comment]
We may be hitting into an [asyncio bug][1] here.  Perhaps a response there will include a workaround or some other advice we can use to resolve this on our end.  IMO this is quite important for us to deal with, as it could lead to a large variety of sporadic failures.  (I suspect it's a core component behind the spread of failures observed in gh-4699 when the config manager is switched to the `WitlessRunner`.)

[1]: https://bugs.python.org/issue41594

**Comment by yarikoptic** [comment]
Eh, that's "exciting".  I will try reproducer later on... Wondering if it is due to short-lived printf or not.

**Comment by kyleam** [comment]
> Wondering if it is due to short-lived printf or not.

Yeah, I _think_ the speed of the program is relevant.  When I tried with the `python -c import datetime; print(datetime.datetime.now())'` [example][1] from the python docs, I couldn't trigger it.  But I could when I switched to `date` or other quicker programs (including git-add).

[1]: https://docs.python.org/3/library/asyncio-protocol.html#loop-subprocess-exec-and-subprocessprotocol
